File: flutter/lib/src/interface/room_close_reason.dart
import '/src/util/rust_handles_storage.dart';
import 'enums.dart' show RoomCloseKind;

export 'enums.dart' show RoomCloseKind;

/// Reason of why a `Room` has been closed.
///
/// This struct is passed into the `RoomHandle.onClose()` callback.
abstract class RoomCloseReason implements SyncPlatformHandle {
  /// Returns a close reason of the `Room`.
  RoomCloseKind reason();

  /// Indicates whether the `Room` was closed by server.
  bool isClosedByServer();
}

File: src/api/dart/api/room_close_reason.rs
//! Reason of a [`Room`] closing.

#[cfg(doc)]
use crate::room::Room;
use crate::{
    api::{RoomCloseKind, dart::api::ForeignClass},
    room as core,
};

/// Reason of why a [`Room`] is closed.
#[derive(Clone, Copy, Debug)]
pub struct RoomCloseReason {
    /// [`Room`]'s close reason.
    pub reason: RoomCloseKind,

    /// Indicator whether the [`Room`] was closed by server.
    pub is_closed_by_server: bool,
}

impl From<core::RoomCloseReasonImpl> for RoomCloseReason {
    fn from(r: core::RoomCloseReasonImpl) -> Self {
        Self { reason: r.reason, is_closed_by_server: r.is_closed_by_server }
    }
}

impl ForeignClass for RoomCloseReason {}
